File: cp_rust/src/bin/cses-1071-number-spiral.rs
// Created by Ayush Biswas at 2025/07/09 22:58
// https://cses.fi/problemset/task/1071
use cp_lib::*;

// @code begin
use cpio::*;

sol_n! {
    fn solution(
        [x, y]: [usize; 2]
    ) -> usize {
        let diagonal = x.max(y);
        let diag_value = 1 + (diagonal - 1) * diagonal;
        if diagonal % 2 == 0 {
            diag_value + x - y
        } else {
            diag_value - x + y
        }
    }
}

// @code end

File: cp_haskell/bin/cses-1071-number-spiral.hs
-- Created by Ayush Biswas at 2026/03/03 18:33
-- https://cses.fi/problemset/task/1071

-- @code begin
import Control.Monad (replicateM)

getInts :: IO [Int]
getInts = map read . words <$> getLine

sol :: [Int] -> Int
sol [x, y]
  | even n = k + x - y
  | odd n = k + y - x
  where
    n = max x y
    k = n ^ 2 - n + 1

main :: IO ()
main = do
  n :: Int <- read <$> getLine
  res <- replicateM n (sol <$> getInts)
  mapM_ print res

-- @code end